File: 실버/1676(팩토리얼 0의 개수)/Main.cpp
#include <iostream>
#include <utility>	// pair 사용
#include <algorithm>	// min 사용
using namespace std;

int main() {
	// IO 속도 향상
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	
	int N;
	cin >> N;

	int cnt[501];
	pair<int, int>fact[501];

	cnt[0] = 0;
	cnt[1] = 0;
	for (int i = 2; i <= N; i++) {
		int temp = i;
		int two = 0;
		int five = 0;

		while (1) {
			if (temp % 2 == 0) {
				two++;
				temp /= 2;
			}
			else
				break;
		}
		temp = i;

		while (1) {
			if (temp % 5 == 0) {
				five++;
				temp /= 5;
			}
			else
				break;
		}

		fact[i].first = fact[i - 1].first + two;
		fact[i].second = fact[i - 1].second + five;
		cnt[i] = min(fact[i].first, fact[i].second);
	}
	cout << cnt[N] << endl;
	return 0;
}

File: 실버/1676(팩토리얼 0의 개수)/Main.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

class Pair{
	int five;
	int two;

	Pair(int five, int two){
		this.five = five;
		this.two = two;
	}
}

class Main {
	public static void main(String[] args) throws IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		
		int[] cnt = new int[N+1];
		Pair[] fact = new Pair[N+1];
		for(int i=0; i<=N; i++)
			fact[i] = new Pair(0,0);

		for(int i=2; i<=N; i++){
			int temp = i;
			int five = 0;
			int two = 0;
			
			// 새로 곱하는 수의 인수 2의 개수 구하기
			while(true){
				if(temp % 2 == 0){
					two++;
					temp /= 2;
				}
				else
					break;
			}
			temp = i;

			// 새로 곱하는 수의 인수 5의 개수 구하기
			while(true){
				if(temp % 5 == 0){
					five++;
					temp /= 5;
				}
				else
					break;
			}
			fact[i].two = fact[i-1].two + two;
			fact[i].five = fact[i-1].five + five;

			cnt[i] = Math.min(fact[i].two, fact[i].five);
		}
		System.out.println(cnt[N]);
		br.close();
	}
}